File: backend/app/schemas/attendance.py
from pydantic import BaseModel, Field
from datetime import date, time
from typing import List, Optional
from app.schemas.subject import SubjectResponse

class LectureOccurrenceResponse(BaseModel):
    id: int
    semester_id: int
    subject_id: int
    date: date
    start_time: time
    end_time: time
    attendance_status: str
    subject: SubjectResponse

    class Config:
        from_attributes = True

class AttendanceUpdate(BaseModel):
    status: str = Field(..., description="'present', 'absent', 'cancelled', 'unmarked'")

class UpcomingDaySchedule(BaseModel):
    date: date
    day_label: str
    event_type: Optional[str] = None
    description: Optional[str] = None
    occurrences: List[LectureOccurrenceResponse]

    class Config:
        from_attributes = True

class SubjectInitialization(BaseModel):
    subject_id: int
    initial_conducted: int = Field(..., ge=0)
    initial_attended: int = Field(..., ge=0)

class AttendanceInitializationRequest(BaseModel):
    initializations: List[SubjectInitialization]

File: frontend/src/services/attendance.ts
import api from "./api";
import type { Subject } from "./subject";

export interface LectureOccurrence {
  id: number;
  semester_id: number;
  subject_id: number;
  date: string; // "YYYY-MM-DD"
  start_time: string; // "HH:MM:SS" or "HH:MM"
  end_time: string;
  attendance_status: "unmarked" | "present" | "absent" | "cancelled";
  subject: Subject;
}

export const attendanceService = {
  async getByDate(semesterId: number, dateStr?: string): Promise<LectureOccurrence[]> {
    const params = dateStr ? { date_query: dateStr } : {};
    const response = await api.get(`/semesters/${semesterId}/attendance`, { params });
    return response.data;
  },

  async getToday(semesterId: number): Promise<LectureOccurrence[]> {
    const response = await api.get(`/semesters/${semesterId}/attendance/today`);
    return response.data;
  },

  async updateStatus(semesterId: number, occurrenceId: number, status: "present" | "absent" | "cancelled" | "unmarked"): Promise<LectureOccurrence> {
    const response = await api.put(`/semesters/${semesterId}/attendance/${occurrenceId}`, { status });
    return response.data;
  },
};
